references/classification/train_quantization.py

```diff
@@ -0,0 +1,254 @@
+from __future__ import print_function
+import datetime
+import os
+import time
+import sys
+import copy
+
+import torch
+import torch.utils.data
+from torch import nn
+import torchvision
+import torch.quantization
+import utils
+from train import train_one_epoch, evaluate, load_data
+
+
+def main(args):
+
+    if args.output_dir:
+        utils.mkdir(args.output_dir)
+
+    utils.init_distributed_mode(args)
+    print(args)
+
+    if args.post_training_quantize and args.distributed:
+        raise RuntimeError("Post training quantization example should not be performed "
+                           "on distributed mode")
+
+    # Set backend engine to ensure that quantized model runs on the correct kernels
+    if args.backend not in torch.backends.quantized.supported_engines:
+        raise RuntimeError("Quantized backend not supported: " + str(args.backend))
+    torch.backends.quantized.engine = args.backend
+
+    device = torch.device(args.device)
+    torch.backends.cudnn.benchmark = True
+
+    # Data loading code
+    print("Loading data")
+    train_dir = os.path.join(args.data_path, 'train')
+    val_dir = os.path.join(args.data_path, 'val')
+
+    dataset, dataset_test, train_sampler, test_sampler = load_data(train_dir, val_dir,
+                                                                   args.cache_dataset, args.distributed)
+    data_loader = torch.utils.data.DataLoader(
+        dataset, batch_size=args.batch_size,
+        sampler=train_sampler, num_workers=args.workers, pin_memory=True)
+
+    data_loader_test = torch.utils.data.DataLoader(
+        dataset_test, batch_size=args.eval_batch_size,
+        sampler=test_sampler, num_workers=args.workers, pin_memory=True)
+
+    print("Creating model", args.model)
+    # when training quantized models, we always start from a pre-trained fp32 reference model
+    model = torchvision.models.quantization.__dict__[args.model](pretrained=True, quantize=args.test_only)
+    model.to(device)
+
+    if not (args.test_only or args.post_training_quantize):
+        model.fuse_model()
+        model.qconfig = torch.quantization.get_default_qat_qconfig(args.backend)
+        torch.quantization.prepare_qat(model, inplace=True)
+
+        optimizer = torch.optim.SGD(
+            model.parameters(), lr=args.lr, momentum=args.momentum,
+            weight_decay=args.weight_decay)
+
+        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer,
+                                                       step_size=args.lr_step_size,
+                                                       gamma=args.lr_gamma)
+
+    criterion = nn.CrossEntropyLoss()
+    model_without_ddp = model
+    if args.distributed:
+        model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[args.gpu])
+        model_without_ddp = model.module
+
+    if args.resume:
+        checkpoint = torch.load(args.resume, map_location='cpu')
```

Comment:
For training, should the device be args.device instead of being fixed to cpu?

Reply:
We force loading to happen on the CPU because this script is supposed to work on machines without GPU as well.

In the (unlikely) case someone has a checkpoint from the internet which was saved on CUDA, and wants to do some training on the CPU (because they don't have GPU), the loading code would fail.
This makes sure that we are safe in this situation.

Note that loading the `checkpoint` on the CPU is always safe, because `load_state_dict` can perform inter-device copies, so the model can be on the GPU and the state-dict on the CPU